src/mono/mono/mini/simd-intrinsics.c

```diff
@@ -3093,16 +3209,113 @@ emit_vector_2_3_4 (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSignature *f
 
 		return min;
 	}
-	case SN_Conjugate:
-	case SN_Distance:
-	case SN_DistanceSquared:
+	case SN_Distance: 
+	case SN_DistanceSquared: {
+#if defined(TARGET_ARM64)
+		MonoInst *diffs = emit_simd_ins_for_sig (cfg, klass, OP_XBINOP, OP_FSUB, MONO_TYPE_R4, fsig, args);
+		MonoInst *diffs_squared = emit_simd_ins (cfg, klass, OP_XBINOP, diffs->dreg, diffs->dreg);
+		diffs_squared->inst_c0 = OP_FMUL;
+		diffs_squared->inst_c1 = MONO_TYPE_R4;
+
+		switch (id) {
+		case SN_Distance:
+			return emit_sum_sqrt_vector_2_3_4 (cfg, klass, diffs_squared);
+		case SN_DistanceSquared:
+			return emit_sum_vector (cfg, fsig->params [0], MONO_TYPE_R4, diffs_squared);
+		default:
+			g_assert_not_reached ();
+		}
+#endif
+	}
+	break;
 	case SN_Length:
-	case SN_LengthSquared:
-	case SN_Lerp:
+	case SN_LengthSquared: {
+#if defined (TARGET_ARM64)
+		int src1 = load_simd_vreg (cfg, cmethod, args [0], NULL);
+
+		MonoInst *vec_squared = emit_simd_ins (cfg, klass, OP_XBINOP, src1, src1);
+		vec_squared->inst_c0 = OP_FMUL;
+		vec_squared->inst_c1 = MONO_TYPE_R4;
+
+		switch (id) {
+		case SN_Length:
+			return emit_sum_sqrt_vector_2_3_4 (cfg, klass, vec_squared);
+		case SN_LengthSquared:
+			return emit_sum_vector (cfg, type, MONO_TYPE_R4, vec_squared);
+		default:
+			g_assert_not_reached ();
+		}
+#endif
+	}
+	break;
+	case SN_Lerp: {
+#if defined (TARGET_ARM64)
+		MonoInst* v1 = args [1];
+		if (!strcmp ("Quaternion", m_class_get_name (klass))) {
```

Comment:
Quaternion.Lerp is not marked as intrinsic in the libraries: https://github.com/dotnet/runtime/blob/db7d26988897328a518229eee49251696b5942f6/src/libraries/System.Private.CoreLib/src/System/Numerics/Quaternion.cs#L493-L498
However, if this implementation generates better codegen we should probably keep it.

Reply:
The intrinsified version is around 40% faster on my machine

Reply:
Why is the intrinsified version faster here? Is it fundamentally doing something differently from the managed implementation or is there potentially a missing JIT optimization?

Reply:
Or perhaps there is simply missing a change on the managed side and so its using scalar logic rather than any actual vectorization and a better fix is to update the managed impl?

-----

We've typically tried to keep a clear separation between `intrinsic` functionality and more complex methods.

APIs like `operator +` or `Sqrt` are generally mapped to exactly 1 hardware instruction and this is the case for most platforms.

APIs like `DotProduct` or even `Create` may be mapped to exactly 1 hardware instruction on some platforms and are fairly "core" to the general throughput considerations of many platforms.

APIs like `Quaternion.Lerp` or `CopyTo` are more complex functions which use multiple instructions on all platforms and which may even require branching or masking logic. So, we've typically tried to keep them in managed and have them use the intrinsic APIs instead.

Reply:
I agree that we should align Mono's behavior with CoreCLR, not intrinsifying `Quaternion.Lerp` or `CopyTo` for mono either.